src/field_theory/separable.lean

```diff
@@ -68,4 +178,148 @@ by { rw [separable_def, derivative_mul], exact ((hf.mul_right h).add_mul_left_ri
+    of_irreducible_map ↑(expand F p) (by rwa ← expand_contract hp H at hf),
+  expand_contract hp H⟩
+else or.inl $ (separable_iff_derivative_ne_zero hf).2 H
+
+theorem exists_separable_of_irreducible {f : polynomial F} (hf : irreducible f) (hf0 : f ≠ 0) :
+  ∃ (n : ℕ) (g : polynomial F), g.separable ∧ expand F (p ^ n) g = f :=
+begin
+  generalize hn : f.nat_degree = N, unfreezingI { revert f },
+  apply nat.strong_induction_on N, intros N ih f hf hf0 hn,
+  rcases separable_or hp hf with h | ⟨h1, g, hg, hgf⟩,
+  { refine ⟨0, f, h, _⟩, rw [nat.pow_zero, expand_one] },
+  { cases N with N,
+    { rw [nat_degree_eq_zero_iff_degree_le_zero, degree_le_zero_iff] at hn,
+      rw [hn, separable_C, is_unit_iff_ne_zero, not_not] at h1,
+      rw [h1, C_0] at hn, exact absurd hn hf0 },
+    have hg1 : g.nat_degree * p = N.succ,
+    { rwa [← nat_degree_expand, hgf] },
+    have hg2 : g.nat_degree ≠ 0,
+    { intro this, rw [this, zero_mul] at hg1, cases hg1 },
+    have hg3 : g.nat_degree < N.succ,
+    { rw [← mul_one g.nat_degree, ← hg1],
+      exact nat.mul_lt_mul_of_pos_left hp.one_lt (nat.pos_of_ne_zero hg2) },
+    have hg4 : g ≠ 0,
+    { rintro rfl, exact hg2 nat_degree_zero },
+    rcases ih _ hg3 hg hg4 rfl with ⟨n, g, hg5, rfl⟩, refine ⟨n+1, g, hg5, _⟩,
+    rw [← hgf, expand_expand, nat.pow_succ, mul_comm] }
+end
+
+theorem of_separable_expand {f : polynomial F} (n : ℕ)
+  (hf : (expand F (p ^ n) f).separable) : is_unit f ∨ n = 0 :=
+begin
+  rw classical.or_iff_not_imp_right, intro hn,
+  have hf2 : (expand F (p ^ n) f).derivative = 0,
+  { by rw [derivative_expand, nat.cast_pow, char_p.cast_eq_zero,
+      zero_pow (nat.pos_of_ne_zero hn), zero_mul, mul_zero] },
+  rw [separable_def, hf2, is_coprime_zero_right, is_unit_iff] at hf, rcases hf with ⟨r, hr, hrf⟩,
+  rw [eq_comm, expand_eq_C (nat.pow_pos hp.pos _)] at hrf,
+  rwa [hrf, is_unit_C]
+end
+
+theorem unique_separable_of_irreducible {f : polynomial F} (hf : irreducible f) (hf0 : f ≠ 0)
+  (n₁ : ℕ) (g₁ : polynomial F) (hg₁ : g₁.separable) (hgf₁ : expand F (p ^ n₁) g₁ = f)
+  (n₂ : ℕ) (g₂ : polynomial F) (hg₂ : g₂.separable) (hgf₂ : expand F (p ^ n₂) g₂ = f) :
+  n₁ = n₂ ∧ g₁ = g₂ :=
+begin
+  revert g₁ g₂, wlog hn : n₁ ≤ n₂ := le_total n₁ n₂ using [n₁ n₂, n₂ n₁] tactic.skip,
+  unfreezingI { intros, rw le_iff_exists_add at hn, rcases hn with ⟨k, rfl⟩,
+    rw [← hgf₁, nat.pow_add, expand_mul, expand_inj (nat.pow_pos hp.pos n₁)] at hgf₂, subst hgf₂,
+    subst hgf₁,
+    rcases of_separable_expand hp k hg₁ with h | rfl,
+    { rw is_unit_iff at h, rcases h with ⟨r, hr, rfl⟩,
+      simp_rw expand_C at hf, exact absurd (is_unit_C.2 hr) hf.1 },
+    { rw [add_zero, nat.pow_zero, expand_one], split; refl } },
+  exact λ g₁ g₂ hg₁ hgf₁ hg₂ hgf₂, let ⟨hn, hg⟩ := this g₂ g₁ hg₂ hgf₂ hg₁ hgf₁ in ⟨hn.symm, hg.symm⟩
+end
+
+end char_p
+
+section char_zero
+
+variables [char_zero F]
+
+theorem separable_of_irreducible {f : polynomial F} (hf : irreducible f) (hf0 : f ≠ 0) :
```

Comment:
```suggestion
theorem irredicuble.separable {f : polynomial F} (hf : irreducible f) (hf0 : f ≠ 0) :
```
Maybe? Feel free to disagree. (I guess this means it has to be moved outside of some namespace...)

Reply:
(note to self)